File: 103-binary-tree-zigzag-level-order-traversal/binary-tree-zigzag-level-order-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        vector<vector<int>> ans;
        if(!root) return ans;
        stack<TreeNode*> s1; //l->r
        stack<TreeNode*> s2; //r->l
        s1.push(root);
        while(!s1.empty() || !s2.empty()){
            

            if(!s1.empty()){
                vector<int> arr;
                while(!s1.empty()){
                    TreeNode* temp=s1.top();
                    s1.pop();
                    arr.push_back(temp->val);
                    if(temp->left) s2.push(temp->left);
                    if(temp->right) s2.push(temp->right);
                }
                ans.push_back(arr);
            }
            if(!s2.empty()){
                vector<int> arr;
                while(!s2.empty()){
                    TreeNode* temp=s2.top();
                    s2.pop();
                    arr.push_back(temp->val);
                    if(temp->right) s1.push(temp->right);
                    if(temp->left) s1.push(temp->left);
                }
                ans.push_back(arr);
            }
            


        }
        return ans;

    }
};

File: 103-binary-tree-zigzag-level-order-traversal/binary-tree-zigzag-level-order-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>> ans= new ArrayList<>();
        Queue<TreeNode> q= new ArrayDeque<>();
        if(root==null) return new ArrayList<>();
        q.offer(root);
        TreeNode curr;
        boolean LeftToRight=true;
        while(!q.isEmpty()){
            int size=q.size();
            
            List<Integer> ll= new ArrayList<>();

            while(size>0){
                curr=q.poll();
                
                    if(LeftToRight){
                        ll.add(curr.val);
                    }
                    else{
                        ll.add(0,curr.val);
                    }
                    
                    if(curr.left!=null) q.offer(curr.left);
                    if(curr.right!=null) q.offer(curr.right);
                   
                
                size--;
            }
             LeftToRight=!LeftToRight;
            ans.add(ll);
        }
        return ans;
    }
}